feat(docs): Better docs for big chunk of schema stuff (#25503)
gatsbyjs/gatsby: docs/docs/schema-generation.md

title: Schema Generation
---

> Once the nodes have been sourced and transformed, the next step is to generate the GraphQL Schema. Gatsby Schema is different from many regular GraphQL schemas is that it's combines plugin or user defined schema information with data inferred from node shapes itself. The latter is called _schema inferrence_. Users or plugin can explicity define whole or part of the schema using [schema customization API](/docs/schema-customization). Usually every node will get a GraphQL Type based on its `node.internal.type` field. When using Schema Customization, all types that implement `Node` inteface become GraphQL Node Types and thus get root level fields for accessing them.

Reviewer: Fix Grammar and Typo

```suggestion
Once the nodes have been sourced and transformed, the next step is to generate the GraphQL Schema. Gatsby Schema is different from many regular GraphQL schemas is that it combines plugin or user defined schema information with data inferred from node shapes itself. The latter is called _schema inferrence_. Users or plugin can explicitly define whole or part of the schema using [schema customization API](/docs/schema-customization). Usually every node will get a GraphQL Type based on its `node.internal.type` field. When using Schema Customization, all types that implement `Node` interface become GraphQL Node Types and thus get root level fields for accessing them.
```

## 3. Legacy schema customization

Before schema customization was added, there were several ways that one could modify the schema. Those were `createNodeField` action, `setFieldsOnGraphQLType` API and `gatsby-config` mappings.

### `createNodeField`

This adds a field under `fields` field (yo dawg). This method is needed because plugins can't modify types that they haven't created, so this is a way to add data to nodes that it doesn't own. This doesn't modify schema directly, instead those fields are picked by inference. There are no plans currently to deprecate this API at the moment.

### `setFieldsOnGraphQLType`

This allows adding GraphQL Fields to any node type. This operates on GraphQL types itself and syntax matches `graphql-js` field definitions. This API will be marked as deprecated in Gatsby v3, moved under flag in Gatsby v4 and removed from Gatsby v5. `createTypes` and `addResolvers` should solve all the use cases when this API is used.

### `graphql-config.js` mapping

> It's possible to connect types throught site configuration using [Node Type Mapping](https://www.gatsbyjs.org/docs/gatsby-config/#mapping-node-types).

Reviewer: Typo

```suggestion
It's possible to connect types throughout site configuration using [Node Type Mapping](https://www.gatsbyjs.org/docs/gatsby-config/#mapping-node-types).
```

> It is a convenient escape hatch and there are no plans to deprecate this API at the moment.

Author: Fixed the phrasing.

Children types are either inferred from data or created using `@childOf` directive, either by parent type name of by mimeType (only for File parent types).

## 5. Processing each type and adding root fields

See [Schema Root Fields and Utility Types](/docs/schema-root-fields) for a more detailed description of this step.

For each type, utility types are created. Those are input object types, used for searching, sorting and filtering, and types for paginated data (Connections and Edges).

For searching and sorting, Gatsby goes through every field in the type and converts them to corresponding Input GraphQL types. Scalars are converted to objects with filter operator fields like "eq" or "ni", ObjectTypes are converted to InputObjectTypes. For sorting, enums are created out of all fields so that one can use that to specify the sort.

Those types are used to create _root fields_ of the schema in `Query` type. For each node type a root field for querying one item and paginated items are created (eg for type BlogPost it would be `blogPost` and `allBlogPost`).

## 6. Merging in third-party schemas

> If a plugin like `gatsby-source-graphql` is used, all third-party schemas that it provided are merged in into the Gatsby schema.

Reviewer: Remove preposition

```suggestion
If a plugin like `gatsby-source-graphql` is used, all third-party schemas that it provided are merged into the Gatsby schema.
```
